### ISSUE TITLE
how to implement SuP

### ISSUE THREAD

**Issue by Shinichi0713** [Shinichi0713/recommendation-ai#8]
はい、PyTorchでのSupConLoss（Supervised Contrastive Loss）の具体的な実装例を紹介します。

---

## 1. シンプルなSupConLoss実装例

以下は、公式論文や有名な実装（[SupContrast公式GitHub](https://github.com/HobbitLong/SupContrast)）を参考にした、バッチ単位でのSupConLossの実装例です。

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class SupConLoss(nn.Module):
    def __init__(self, temperature=0.07):
        super(SupConLoss, self).__init__()
        self.temperature = temperature

    def forward(self, features, labels):
        """
        features: [batch_size, feature_dim]
        labels:   [batch_size]
        """
        device = features.device
        batch_size = features.shape[0]

        # 特徴ベクトルを正規化
        features = F.normalize(features, dim=1)
        
        # 類似度行列を計算
        similarity_matrix = torch.matmul(features, features.T) / self.temperature  # [B, B]

        # マスクを作成（同じクラスなら1, それ以外は0）
        labels = labels.contiguous().view(-1, 1)
        mask = torch.eq(labels, labels.T).float().to(device)  # [B, B]

        # 自分自身は除外
        logits_mask = torch.ones_like(mask) - torch.eye(batch_size).to(device)
        mask = mask * logits_mask

        # 各サンプルについて、正例のみに対する損失を計算
        exp_sim = torch.exp(similarity_matrix) * logits_mask
        log_prob = similarity_matrix - torch.log(exp_sim.sum(dim=1, keepdim=True) + 1e-12)

        # 正例ペアごとに平均
        mean_log_prob_pos = (mask * log_prob).sum(dim=1) / (mask.sum(dim=1) + 1e-12)

        # 損失の平均（負号をつけて最小化）
        loss = -mean_log_prob_pos.mean()
        return loss
```

---

## 2. 使い方例

```python
# バッチサイズ32、特徴次元128と仮定
features = torch.randn(32, 128)
labels = torch.randint(0, 10, (32,))

criterion = SupConLoss(temperature=0.07)
loss = criterion(features, labels)
print(loss)
```

---

## 3. ポイント

- **temperature**  
  コントラスト学習でよく使われる温度パラメータです（小さいほど鋭い分布）。
- **特徴正規化**  
  コサイン類似度ベースのため、特徴ベクトルは正規化します。
- **同じクラス同士のペアのみを「正例」として損失を計算**します。

---

### 参考リンク
- [SupContrast公式実装（PyTorch）](https://github.com/HobbitLong/SupContrast/blob/master/losses.py)
- [論文: Supervised Contrastive Learning](https://arxiv.org/abs/2004.11362)

---

ご参考になれば幸いです。  
以上です。
